scripts/srcinfo.sh

```diff
@@ -20,28 +20,119 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with Pacstall. If not, see <https://www.gnu.org/licenses/>.
+#
+# based on makepkg's pkgbuild.sh,
+# Copyright (C) 2009-2024 Pacman Development Team <pacman-dev@lists.archlinux.org>
 
 shopt -s extglob
 
-function die() {
-  printf >&2 'ERROR: %s\n' "$@"
-  exit 1
+array_build() {
+  local dest="${1}" src="${2}" i keys values
+  declare -p "$2" &> /dev/null || return 1
+  eval "keys=(\"\${!$2[@]}\")"
+  eval "${dest}=()"
+  for i in "${keys[@]}"; do
+    values+=("printf -v '${dest}[${i}]' %s \"\${${src}[${i}]}\";")
+  done
+  eval "${values[*]}"
+}
+
+extract_global_variable() {
+  local attr="${1}" isarray="${2}" outputvar="${3}" ref
+  if ((isarray)); then
+    array_build ref "${attr}"
+    ((${#ref[@]})) && array_build "${outputvar}" "${attr}"
+  else
+    [[ -n ${!attr} ]] && printf -v "${outputvar}" %s "${!attr}"
+  fi
+}
+
+extract_function_variable() {
+  local funcname="${1}" attr="${2}" isarray="${3}" outputvar="${4}"
+  local attr_regex decl r=1
+  if ((isarray)); then
+    printf -v attr_regex '^[[:space:]]*%s[[:space:]]*\\(' "${attr}"
+  else
+    printf -v attr_regex '^[[:space:]]*%s[[:space:]]*=' "${attr}"
+  fi
+  local func_body
+  func_body=$(declare -f "${funcname}" 2> /dev/null)
+  [[ -z ${func_body} ]] && return 1
+  IFS=$'\n' read -r -d '' -a lines <<< "${func_body}"
+  for line in "${lines[@]}"; do
+    [[ ${line} =~ ${attr_regex} ]] || continue
+    decl=${line##*([[:space:]])}
+    eval "${decl/#${attr}/${outputvar}}"
+    r=0
+  done
+  return "${r}"
+}
+
+get_pkgbuild_attribute() {
+  local pkgname="${1}" attrname="${2}" isarray="${3}" outputvar="${4}"
+  if ((isarray)); then
+    eval "${outputvar}=()"
+  else
+    printf -v "${outputvar}" %s ''
+  fi
+  if [[ -n ${pkgname} ]]; then
+    extract_global_variable "${attrname}" "${isarray}" "${outputvar}"
+    extract_function_variable "package_${pkgname}" "${attrname}" "${isarray}" "${outputvar}"
+  else
+    extract_global_variable "${attrname}" "${isarray}" "${outputvar}"
+  fi
+}
+
+srcinfo_write_attr() {
+  local attrname="${1}" attrvalues=("${@:2}")
+  attrvalues=("${attrvalues[@]//+([[:space:]])/ }")
+  attrvalues=("${attrvalues[@]#[[:space:]]}")
+  attrvalues=("${attrvalues[@]%[[:space:]]}")
+  printf "\t${attrname} = %s\n" "${attrvalues[@]}"
+}
+
+pkgbuild_extract_to_srcinfo() {
+  local pkgname="${1}" attrname=$2 isarray=$3 outvalue
+  if get_pkgbuild_attribute "${pkgname}" "${attrname}" "${isarray}" 'outvalue'; then
+    srcinfo_write_attr "${attrname}" "${outvalue[@]}"
+  fi
+}
+
+srcinfo_write_section_details() {
+  local attr package_arch a
+  for attr in "${singlevalued[@]}"; do
+    pkgbuild_extract_to_srcinfo "$1" "${attr}" 0
+  done
+
+  for attr in "${multivalued[@]}"; do
+    pkgbuild_extract_to_srcinfo "$1" "${attr}" 1
+  done
+
+  get_pkgbuild_attribute "$1" 'arch' 1 'package_arch'
+  for a in "${package_arch[@]}"; do
+    [[ ${a} == any || ${a} == all ]] && continue
```

Comment:
So are we supporting `any` / `all` as an `arch` entry now? If so, is it not implicit anymore?

Reply:
We already do support them, and they're still implicit. This is just saying if arch=any or all, don't try to append arch enhanced arrays